### PR TITLE
Fix start time calculation in schedule spec

### PR DESCRIPTION
**Pull request by dnr**
## What changed?
- Fix handling of start time in schedule specs that could result in GetNextTime returning the time that it was given.
- Introduce a SpecVersion as part of tweakables but separate from workflow logic version to control this behavior.
- Non-workflow logic can always use the latest version.
- Workflow logic is initialized with the fixed version, which is technically not downgrade-compatible, but it should affect very few workflows.

## Why?
This would lead to incorrect results and infinite loops.

## How did you test it?
new unit test, manual replay tests

## Potential risks
This was done in a way to be upgrade-compatible, but is not downgrade-compatible for workflows that hit this edge case after the upgrade.

### CODE REVIEW

File: service/worker/scheduler/spec.go [hunk 1]
```diff
@@ -295,9 +296,7 @@ func (cs *CompiledSpec) CanonicalForm() *schedulepb.ScheduleSpec {
 func (cs *CompiledSpec) GetNextTime(jitterSeed string, after time.Time) GetNextTimeResult {
 	// If we're starting before the schedule's allowed time range, jump up to right before
 	// it (so that we can still return the first second of the range if it happens to match).
-	if cs.spec.StartTime != nil && after.Before(timestamp.TimeValue(cs.spec.StartTime)) {
-		after = cs.spec.StartTime.AsTime().Add(-time.Second)
-	}
+	after = util.MaxTime(after, cs.spec.StartTime.AsTime().Add(-time.Second))
```
Comment (yiminc) [inline service/worker/scheduler/spec.go:299]: is there risk of NPE on cs.spec.StartTime?

Comment (dnr) [reply service/worker/scheduler/spec.go:299]: relying on AsTime to return unix epoch, which is way in the past. does it need a comment?

File: service/worker/scheduler/workflow.go [hunk 2]
```diff
@@ -175,6 +175,7 @@ type (
 		ReuseTimer                        bool                     // Whether to reuse timer. Used for workflow compatibility.
 		NextTimeCacheV2Size               int                      // Size of next time cache (v2)
 		SpecFieldLengthLimit              int                      // item limit per spec field on the ScheduleInfo memo
+		SpecVersion                       SpecVersion              // version for spec logic
```
Comment (yiminc) [inline service/worker/scheduler/workflow.go:178]: Just curious, why not just a simple bool?

**Review by yiminc — APPROVED** [review]

File: service/worker/scheduler/spec_test.go [hunk 3]
```diff
@@ -407,6 +407,20 @@ func (s *specSuite) TestSpecStartTime() {
 	)
 }
 
+func (s *specSuite) TestSpecStartTimeMinusOneSecond() {
```
Comment (lina-temporal) [inline service/worker/scheduler/spec_test.go:410]: (minor) Maybe a comment to call out that the test case validates the `FixStartTimeBug` version case?

**Review by lina-temporal — APPROVED** [review]

File: service/worker/scheduler/workflow.go (hunk 2 above)
Comment (dnr) [reply service/worker/scheduler/workflow.go:178]: Basically the same reason for `Version`: if we added a bool for each fix, the struct would get bigger and bigger. This way we can just increment a number.
